lib/namespace-hmac.js

```diff
@@ -0,0 +1,289 @@
+const crypto = require('crypto');
+const database = require('bedrock-mongodb');
+const jwt = require('jsonwebtoken');
+const BedrockError = bedrock.util.BedrockError;
+const Namespace = require('./namespace');
+
+const logger = bedrock.loggers.get('app');
+
+class NamespaceHmac extends Namespace {
+  constructor() {
+    super();
+    this._keyState = null;
+  }
+
+  createState(options, callback) {
+    const opts = _.assign({}, options, {provision: true});
+    this._getState(opts, err => {
+      if(err) {
+        return callback(err);
+      }
+      callback(null, options);
+    });
+  }
+
+  getKey(options, callback) {
+    this._getState(options, (err, result) => {
+      if(err) {
+        return callback(err);
+      }
+      callback(null, {id: result.key.id, material: result.key.data});
+    });
+  }
+
+  /**
+   * Verify a JWT.
+   *
+   * @param token the token to be verified.
+   * @param options the options to use:
+   *   algorithms an array of acceptable HMAC signature algorithms.
+   *   clockToleranceInSecs the acceptable clock skew.
+   *
+   * @param callback(err, tokenPayload) called once the operation completes.
+   */
+  verify(token, options, callback) {
+    let decodedToken;
+    let header;
+    try {
+      decodedToken = jwt.decode(token, {complete: true});
+      header = decodedToken.header;
+    } catch(e) {
+      return callback(e);
+    }
+
+    async.auto({
+      get: callback => this._getState(callback),
+      verify: ['get', (callback, results) => {
+        const state = results.get;
+        // get key that matches key ID
+        let key;
+        if(header.kid === state.key.id) {
+          key = state.key;
+        } else if(state.previousKey && header.kid === state.previousKey.id) {
+          key = state.previousKey;
+        } else {
+          return callback(new Error('Invalid key identifier in token.'));
+        }
+        // decode and verify the token
+        jwt.verify(
+          token, key.data, {
+            algorithms: options.verifyAlgorithms,
+            clockTolerance: options.clockToleranceInSecs
+          }, callback);
+      }]
+    }, (err, results) => callback(err, results ? results.verify : false));
+  }
+
+  _get(id, callback) {
+    database.collections.jwtKeyStoreHmac.findOne(
+      {id: database.hash(id)}, {state: true}, (err, record) => {
+        if(err) {
+          logger.error('Mongo error when trying to find key state.', err);
+          return callback(new BedrockError(
+            'Failed to find key state due to internal error.',
+            'InternalError',
+            {key: id, httpStatusCode: 500, 'public': true}, err));
+        }
+        if(!record) {
+          return callback(new BedrockError(
+            'Key not found.',
+            'NotFound',
+            {key: id, httpStatusCode: 404, 'public': true}));
+        }
+        callback(null, record.state, record.meta);
+      });
+  }
+
+  _insert(id, state, callback) {
```

Comment:
It seems like the `insert` was handled in the base class and isn't necessary here. It's also going into a separate collection which I don't understand -- we can chat about it.